File: Google Code Jam/2019/Round 2/new_elements_part_2.py
# @Author: Joey Teng
# @Email:  joey.teng.dev@gmail.com
# @Filename: new_elements_part_2.py
# @Last modified by:   Joey Teng
# @Last modified time: 19-May-2019
import fractions

T = int(input())
for t in range(1, T + 1):
    impossible = 0
    N = int(input())
    molecule = []
    for i in range(N):
        molecule.append(tuple(map(int, str(input()).split(' '))))
    fraction = []
    for i in range(1, N):
        x = molecule[i - 1][0] - molecule[i][0]
        y = molecule[i][1] - molecule[i - 1][1]
        if x == 0 and y < 0 or y == 0 and x > 0:
            print("Case #{}: IMPOSSIBLE".format(t))
            impossible = 1
            break
        if x == 0 or y == 0:
            continue

        _fraction = fractions.Fraction(x, y)
        if x < 0 and _fraction < 0:
            continue

        if x < 0:
            fraction.append((0, fractions.Fraction(x, y)))
        else:
            fraction.append((1, fractions.Fraction(x, y)))

    if impossible:
        continue

    if len(fraction) == 0:
        x = 1
        y = 1
        print("Case #{0}: {1} {2}".format(t, x, y))
        continue

    fraction.sort()
    if fraction[0][0] == 0:
        checking = fraction[0][1]
        for sign, frac in fraction:
            if sign == 1:
                if frac <= checking:
                    print("Case #{}: IMPOSSIBLE".format(t))
                    impossible = 1
                break

    if impossible:
        continue

    x = 0
    y = 0

    print("Case #{0}: {1} {2}".format(t, x, y))

File: Google Code Jam/2019/Round 2/new_elements_part_2.cpp
/*
 * @Author: Joey Teng
 * @Email:  joey.teng.dev@gmail.com
 * @Filename: new_elements_part_2.cpp
 * @Last modified by:   Joey Teng
 * @Last modified time: 19-May-2019
 */
#include <bits/stdc++.h>

int64_t gcd(int64_t a, int64_t b) {
  if (b == 0) {
    return a;
  }
  return gcd(b, a % b);
}

int64_t cmp(std::pair<int64_t, int64_t> a, std::pair<int64_t, int64_t> b) {
  if (a.first * b.second < a.second * b.first) {
    return true;
  }
  return false;
}

int main() {
  int T = 0;
  std::cin >> T;
  for (int t = 1; t <= T; ++t) {
    int N = 0;
    std::cin >> N;
    std::vector<std::pair<int64_t, int64_t>> molecule;
    std::vector<std::pair<int64_t, int64_t>> inequality;
    for (int i = 0; i < N; ++i) {
      int64_t c = 0;
      int64_t j = 0;
      std::cin >> c >> j;
      molecule.push_back(std::make_pair(c, j));
    }
    for (int i = 1; i < N; ++i) {
      // ax + by < 0
      inequality.push_back(
          std::make_pair(molecule[i - 1].first - molecule[i].first,
                         molecule[i].second - molecule[i - 1].second));
      int64_t _gcd =
          std::abs(gcd(inequality.back().first, inequality.back().second));
      if (_gcd != 1) {
        inequality.back().first /= _gcd;
        inequality.back().second /= _gcd;
      }
    }
    int IMPOSSIBLE = 0;
    // std::cout << std::endl;
    for (auto p : inequality) {
      //   std::cout << ' ' << p.first << ' ' << p.second << std::endl;
      if ((p.first == 0 && p.second < 0) || (p.first > 0 && p.second == 0)) {
        IMPOSSIBLE = 1;
        break;
      }
    }
    std::pair<int64_t, int64_t> checking_negative_pair = std::make_pair(0, 1);
    for (auto p : inequality) {
      if (cmp(p, checking_negative_pair)) {
        checking_negative_pair = p;
      }
    }
    for (auto p : inequality) {
      if (p != checking_negative_pair &&
          (cmp(p, checking_negative_pair) ||
           (p.first * checking_negative_pair.second ==
            p.second * checking_negative_pair.first))) {
        IMPOSSIBLE = 1;
        break;
      }
    }
    if (IMPOSSIBLE == 1) {
      std::cout << "Case #" << t << ": "
                << "IMPOSSIBLE" << std::endl;
      continue;
    }

    std::sort(inequality.begin(), inequality.end(), cmp);

    int64_t _min = INT_MAX;
    int index = -1;
    for (int i = N - 2; i > -1; --i) {
      if (inequality[i].first == 0 || inequality[i].second == 0) {
        continue;
      }
      if (_min > std::abs(inequality[i].first)) {
        _min = std::abs(inequality[i].first);
        index = i;
      }
    }
    // std::cout << index << std::endl;
    int64_t c = 0;
    int64_t j = 0;
    if (index == -1) {
      c = j = 1;
    } else {
      c = inequality[index].first;
      j = inequality[index].second;
      int64_t _gcd = gcd(c, j);
      //   std::cout << "c j " << c << ' ' << j << std::endl;
      c /= _gcd;
      j /= _gcd;
      if (_gcd > 0) {
        ++j;
      } else {
        ++c;
      }
    }

    std::cout << "Case #" << t << ": " << c << ' ' << j << std::endl;
  }

  return 0;
}
